src/payment/unified_qr.rs

```diff
@@ -0,0 +1,424 @@
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Holds a payment handler allowing to create [BIP 21] URIs with an on-chain, [BOLT 11], and [BOLT 12] payment
+//! options.
+//!
+//! [BIP 21]: https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki
+//! [BOLT 11]: https://github.com/lightning/bolts/blob/master/11-payment-encoding.md
+//! [BOLT 12]: https://github.com/lightning/bolts/blob/master/12-offer-encoding.md
+use crate::error::Error;
+use crate::logger::{log_error, FilesystemLogger, Logger};
+use crate::payment::{Bolt11Payment, Bolt12Payment, OnchainPayment};
+use crate::Config;
+
+use lightning::ln::channelmanager::PaymentId;
+use lightning::offers::offer::Offer;
+use lightning_invoice::Bolt11Invoice;
+
+use bip21::de::ParamKind;
+use bip21::{DeserializationError, DeserializeParams, Param, SerializeParams};
+use bitcoin::address::{NetworkChecked, NetworkUnchecked};
+use bitcoin::{Amount, Txid};
+
+use std::sync::Arc;
+use std::vec::IntoIter;
+
+type Uri<'a> = bip21::Uri<'a, NetworkChecked, Extras>;
+
+#[derive(Debug, Clone)]
+struct Extras {
+	bolt11_invoice: Option<Bolt11Invoice>,
+	bolt12_offer: Option<Offer>,
+}
+
+/// A payment handler allowing to create [BIP 21] URIs with an on-chain, [BOLT 11], and [BOLT 12] payment
+/// option.
+///
+/// Should be retrieved by calling [`Node::unified_qr_payment`]
+///
+/// [BIP 21]: https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki
+/// [BOLT 11]: https://github.com/lightning/bolts/blob/master/11-payment-encoding.md
+/// [BOLT 12]: https://github.com/lightning/bolts/blob/master/12-offer-encoding.md
+/// [`Node::unified_qr_payment`]: crate::Node::unified_qr_payment
+pub struct UnifiedQrPayment {
+	onchain_payment: Arc<OnchainPayment>,
+	bolt11_invoice: Arc<Bolt11Payment>,
+	bolt12_payment: Arc<Bolt12Payment>,
+	config: Arc<Config>,
+	logger: Arc<FilesystemLogger>,
+}
+
+impl UnifiedQrPayment {
+	pub(crate) fn new(
+		onchain_payment: Arc<OnchainPayment>, bolt11_invoice: Arc<Bolt11Payment>,
+		bolt12_payment: Arc<Bolt12Payment>, config: Arc<Config>, logger: Arc<FilesystemLogger>,
+	) -> Self {
+		Self { onchain_payment, bolt11_invoice, bolt12_payment, config, logger }
+	}
+
+	/// Generates a URI with an on-chain address, [BOLT 11] invoice and [BOLT 12] offer.
+	///
+	/// The URI allows users to send the payment request allowing the wallet to decide
+	/// which payment method to use. This enables a fallback mechanism: older wallets
+	/// can always pay using the provided on-chain address, while newer wallets will
+	/// typically opt to use the provided BOLT11 invoice or BOLT12 offer.
+	///
+	/// # Parameters
+	/// - `amount_sats`: The amount to be received, specified in satoshis.
+	/// - `message`: A description or note associated with the payment.
+	///   This message is visible to the payee and can provide context or details about the payment.
```

Comment:
> Is it? I think neither the BOLT11 / BOLT12 descriptions nor the BIP21 message are ever sent to the _payee_, but will be seen by the payer, no?
> 
> Also, should this parameter rather be called description to make it less confusing for Lightning users?

No, you're right! I meant payer. I'll update and use description since it's used elsewhere. Thanks for pointing that out!